python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -453,33 +425,61 @@ def _deblend(self, difference, positiveFootprints, negativeFootprints):
-            """Deblend a positive or negative footprint set,
-            and return the deblended children.
-            """
-            sources = afwTable.SourceCatalog(self.schema)
-            footprints.makeSources(sources)
-            self.deblend.run(exposure=difference, sources=sources)
-            self.setPrimaryFlags.run(sources)
-            children = sources["detect_isDeblendedSource"] == 1
-            sources = sources[children].copy(deep=True)
-            # Clear parents, so that measurement plugins behave correctly.
-            sources['parent'] = 0
-            return sources.copy(deep=True)
-
-        positives = deblend(positiveFootprints)
-        negatives = deblend(negativeFootprints)
-
-        sources = afwTable.SourceCatalog(self.schema)
-        sources.reserve(len(positives) + len(negatives))
-        sources.extend(positives, deep=True)
-        sources.extend(negatives, deep=True)
-        return sources, makeFootprints(positives), makeFootprints(negatives)
+        # Merge the positive and negative footprints, but don't grow them.
+        merged_footprints = positiveFootprints
+        merged_footprints.merge(negativeFootprints, 0, 0, False)
+
+        # Create a source catalog from the footprints.
+        table = afwTable.SourceTable.make(self.schema, idFactory)
+        sources = afwTable.SourceCatalog(table)
+        merged_footprints.makeSources(sources)
+
+        # Sky sources must be added before deblending, otherwise the
+        # sources with parent == 0 will be out of order and SFM cannot run.
+        if self.config.doSkySources:
+            self.addSkySources(sources, difference.mask, difference.info.id)
+
+        # Find the footprints with multiple positive peaks no negative peaks.
+        footprints = [src.getFootprint() for src in sources]
+        nPeaks = np.array([len(fp.peaks) for fp in footprints])
+        blend_ids = []
+        skipped_ids = []
+        for src in sources[nPeaks > 1]:
+            peaks = src.getFootprint().peaks
+            peak_flux = np.array([peak.getPeakValue() for peak in peaks])
+            if np.all(peak_flux > 0):
+                blend_ids.append(src.getId())
+            else:
+                skipped_ids.append(src.getId())
+
+        # Deblend the footprints that can be deblended with SourceDeblendTask.
+        temp_cat = afwTable.SourceCatalog(sources.table)
+        for sid in blend_ids:
+            temp_cat.append(sources.find(sid))
+        first_child_index = len(temp_cat)
+        self.deblend.run(exposure=difference, sources=temp_cat)
+
+        # Update the deblended parents and add their children
+        # to the sources catalog.
+        sources.extend(temp_cat[first_child_index:], deep=False)
+
+        # Set detection and primary flags
+        for sid in skipped_ids:
+            src = sources.find(sid)
+            self.deblend.skipParent(src, difference.mask)
+        self.setPrimaryFlags.run(sources)
```

Comment:
I understand what setPrimaryFlags is doing here, but I'm not familiar with `skipParent`. Could you please expand the comment above, since I don't think we're setting any "flags" here, but rather the "not deblended" mask plane (I think?)?

Reply:
Sure. The basic idea is that the deblender already has a method to set all of the appropriate flags for a parent source when deblending is skipped, so we just call that method to set those flags here.

Reply:
I think calling it "detection" as the comment does isn't quite correct, though, since we've already done detection. Maybe `Set blending-related mask planes and flags on the non-deblended parents.` would be more clear?

Reply:
Yes, see my latest commit. But for the `setPrimary` section I still use this comment because technically all of the flags set by `isPrimary` are `detect_<flagName>`, since they are about detection.